File: Data Structures & Algorithms/partition-equal-subset-sum/submission-4.java
class Solution {
     public boolean canPartition(int[] nums) {
        int totalSum = Arrays.stream(nums).sum();
        if (totalSum % 2 != 0) return false;
        int target = totalSum / 2;
//        int[][] cache = new int[nums.length][target + 1];
//        return dfs(0, nums, target, cache);
        
        return bottomUpHelper(nums,target);

    }

    public boolean bottomUpHelper(int[] nums, int target) {
        boolean[] prev = new boolean[target + 1];
        prev[0] = true;

        for (int i = 0; i < nums.length; i++) {
            boolean[] dp = new boolean[target + 1];
            for (int j = 1; j <= target; j++) {
                dp[j] = prev[j] || (j - nums[i] >= 0 && prev[j - nums[i]]);
            }
            prev = dp;

        }

        return prev[target];


    }
}

File: Data Structures & Algorithms/count-subsequences/submission-4.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:

        N = len(s)
        M = len(t)

        prev = [1] * (N+1)

        for i in range(1,M+1):
            dp = [0] * (N+1)
            for j in range(1,N+1):
                skip = dp[j-1]
                include = prev[j-1] if t[i-1]==s[j-1] else 0
                dp[j] = skip+include
            prev=dp
        return prev[-1]
        
        # def dfs(i1: int,i2:int) -> int:
        #     if i2==len(t):
        #         return 1
        #     if i1==len(s):
        #         return 0
        #     if (i1,i2) in cache:
        #         return cache[(i1,i2)]
            
        #     skip = dfs(i1+1,i2)

        #     include = 0
        #     if s[i1]==t[i2]:
        #         include = dfs(i1+1,i2+1)
            
        #     cache[(i1,i2)] =  skip+include
        #     return cache[(i1,i2)]
        
        # return dfs(0,0)